src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp

```diff
@@ -424,22 +424,35 @@ free_unmap_bo(xclDeviceHandle handle, xclBufferHandle boh, void * boptr, size_t
  * helper function for P2P test
  */
 static bool
-p2ptest_set_or_cmp(char *boptr, size_t size, char pattern, bool set)
+p2ptest_set_or_cmp(char *boptr, size_t size, char *valid_data, size_t valid_data_size, bool set)
 {
   int stride = xrt_core::getpagesize();
 
   assert((size % stride) == 0);
+  assert(size > valid_data_size);
   for (size_t i = 0; i < size; i += stride) {
-    if (set) {
-      boptr[i] = pattern;
-    }
-    else if (boptr[i] != pattern) {
-      return false;
-    }
+    if (set)
+      std::memcpy(&(boptr[i]), valid_data, valid_data_size);
+    else if (std::memcmp(&(boptr[i]), valid_data, valid_data_size) != 0) // Verify the written bytes
+        return false;
   }
   return true;
 }
 
+/*
+ * helper function for P2P test
+ */
+static bool
+p2ptest_set_or_cmp(char *boptr, size_t size, char pattern, bool set)
+{
+
+  // Generate the valid data vector using the given pattern
+  std::vector<char> valid_data;
+  valid_data.push_back(pattern);
+
+  return p2ptest_set_or_cmp(boptr, size, valid_data.data(), 1, set);
```

Comment:
Magic numbers are bad.
```suggestion
  return p2ptest_set_or_cmp(boptr, size, valid_data.data(), valid_data.size(), set);
```